src/libasr/pass/intrinsic_functions.h

```diff
@@ -3344,13 +3345,62 @@ namespace Poppar {
         */
         ASR::expr_t *func_call_poppar =Popcnt::POPCNT(b, args[0], return_type, scope);
         body.push_back(al, b.Assignment(result, Mod::MOD(b, func_call_poppar, b.i_t(2, return_type), scope)));
-        ASR::symbol_t *f_sym = make_ASR_Function_t(fn_name, fn_symtab, dep, args, body, result, ASR::abiType::Source, ASR::deftypeType::Implementation, nullptr);
+        ASR::symbol_t *f_sym = make_ASR_Function_t(fn_name, fn_symtab, dep, args,
+            body, result, ASR::abiType::Source, ASR::deftypeType::Implementation, nullptr);
         scope->add_symbol(fn_name, f_sym);
         return b.Call(f_sym, new_args, return_type, nullptr);
     }
 
 } // namespace Poppar
 
+namespace Real {
+
+    static ASR::expr_t *eval_Real(Allocator &al, const Location &loc,
+            ASR::ttype_t* t1, Vec<ASR::expr_t*> &args, diag::Diagnostics& diag) {
+        if (ASR::is_a<ASR::IntegerConstant_t>(*args[0])) {
+            int64_t i = ASR::down_cast<ASR::IntegerConstant_t>(ASRUtils::expr_value(args[0]))->m_n;
+            return make_ConstantWithType(make_RealConstant_t, static_cast<double>(i), t1, loc);
+        } else if (ASR::is_a<ASR::RealConstant_t>(*args[0])) {
+            ASR::RealConstant_t *r = ASR::down_cast<ASR::RealConstant_t>(ASRUtils::expr_value(args[0]));
+            return make_ConstantWithType(make_RealConstant_t, r->m_r, t1, loc);
+        } else if (ASR::is_a<ASR::ComplexConstant_t>(*args[0])) {
+            ASR::ComplexConstant_t *c = ASR::down_cast<ASR::ComplexConstant_t>(ASRUtils::expr_value(args[0]));
+            return make_ConstantWithType(make_RealConstant_t, c->m_re, t1, loc);
+        } else {
+            append_error(diag, "Invalid argument to `real` intrinsic", loc);
+            return nullptr;
+        }
+    }
+
+    static inline ASR::expr_t* instantiate_Real(Allocator &al, const Location &loc,
+        SymbolTable *scope, Vec<ASR::ttype_t*>& arg_types, ASR::ttype_t *return_type,
+        Vec<ASR::call_arg_t>& new_args, int64_t /*overload_id*/) {
+        declare_basic_variables("_lcompilers_real_" + type_to_str_python(arg_types[0]));
+        fill_func_arg("a", arg_types[0]);
+        auto result = declare(fn_name, return_type, ReturnVar);
+        /*
+        function real(a) result(result)
+            real :: a
+            real :: result
+            result = a
+        end function
+        */
+        if (ASR::is_a<ASR::Integer_t>(*arg_types[0])) {
+            body.push_back(al, b.Assignment(result, b.i2r_t(args[0], return_type)));
+        } else if (ASR::is_a<ASR::Real_t>(*arg_types[0])) {
+            body.push_back(al, b.Assignment(result, b.r2r_t(args[0], return_type)));
+        } else if (ASR::is_a<ASR::Complex_t>(*arg_types[0])) {
+            body.push_back(al, b.Assignment(result, EXPR(ASR::make_ComplexRe_t(al, loc,
+                args[0], return_type, nullptr))));
```

Comment:
I think the return type of `make_ComplexRe_t` needs to be updated here. Also, I think we might need to cast the obtained value from `make_ComplexRe_t` to the type of `result` before assignment.

So, it might look like the following:
```suggestion
            body.push_back(al, b.Assignment(result, Cast_to_Result_type(EXPR(ASR::make_ComplexRe_t(al, loc,
                args[0], make_Real(kind=extract_kind_from_expr(args[0))), nullptr))));
```

Please update with appropriate function names.

Reply:
If we cast it to result type, that would mean the same, right? As it was earlier.
(since result has type as return_type)